ncm-spma/src/main/perl/spma/dnf.pm

```diff
@@ -0,0 +1,598 @@
+    my $repoquery_list = Set::Scalar->new( reverse(split ( /\n/, $repodata_rpms)) );
+    my $repoquery_list_noepoch = Set::Scalar->new();
+    my $locked_found           = Set::Scalar->new();
+    my $locked_found_noepoch   = Set::Scalar->new();
+    $self->info( "(" . $repoquery_list->size . " total packages)" );
+    while ( defined( my $p = $repoquery_list->each ) ) {
+        my $t = $p;
+        $t =~ s/^.*://;
+        if ( !$repoquery_list_noepoch->has($t) ) { $repoquery_list_noepoch->insert($t); }
+        if ( $wanted_pkgs_locked->has($t) ) {
+            if ( !$locked_found->has($p) ) {
+                $self->verbose("Found package $p");
+                $locked_found->insert($p);
+                $locked_found_noepoch->insert($t);
+            }
+        }
+    }
+    {
+        my $fh = CAF::FileWriter->new( DNF_PACKAGE_LIST, log => $self );
+        print $fh join ( "\n", @$locked_found );
+        $fh->close();
+        if ( $wanted_pkgs_locked->size != $locked_found->size ) {
+            $self->error( "Version-locked packages are missing from repositories - expected ", $wanted_pkgs_locked->size, ", available ", $locked_found->size, "\n",
+                          "Missing packages: ", $wanted_pkgs_locked - $locked_found_noepoch );
+            return 0;
+        } else {
+            $self->info("all version locked packages available in repositories");
+        }
+    }
+
+    # Test also whether version unlocked packages are present in repositories.
+    {
+        my $found = Set::Scalar->new();
+        while ( defined( my $r = $repoquery_list->each ) ) {
+            my $t = $r;
+            $t =~ s/^.*://;
+            my $name = $t;
+            my $arch = substr($name, rindex($t, '.'));
+            while( (my $end = rindex($name, '-')) != -1) {
+                $name = substr($t, 0, $end);
+                if ( $wanted_pkgs->has("$name$arch") ) {
+                    $found->insert("$name$arch");
+                    last;
+                }
+                if ( $wanted_pkgs->has("$name") ) {
+                    $found->insert("$name");
+                    last;
+                }
+            }
+        }
+        if ( $found->size != $wanted_pkgs->size ) {
+            $self->error("Requested packages are missing from repositories.");
+            $self->error("Missing packages: ", $wanted_pkgs - $found);
+            return 0;
+        }
+    }
+
+    # Continue only if package content is supposed to be changed
+    return 1 unless $t->{run};
+
+    # Run test transaction to get complete list of packages to be present on the system
+    $self->execute_command( [ "rm -rf " . DNF_TEST_CHROOT ], "cleaning DNF test chroot", 1 );
+    $self->execute_command( [ "mkdir -p " . DNF_TEST_CHROOT . "/var/cache" ],                 "setting up DNF test chroot",    1 );
+    $self->execute_command( [ "ln -s /var/cache/dnf " . DNF_TEST_CHROOT . "/var/cache/dnf" ], "setting DNF test chroot cache", 1 );
```

Comment:
These commands feel like they should be defined `Readonly` at the top of the module.